File: libs/cache/cache-simple/src/main/java/org/toolkit4J/libs/cache/simple/SimpleCache.java
package org.toolkit4J.libs.cache.simple;

import lombok.ToString;
import lombok.val;
import net.jodah.expiringmap.ExpirationPolicy;
import net.jodah.expiringmap.ExpiringMap;
import org.jetbrains.annotations.NotNull;
import org.toolkit4J.standard.cache.base.Cache;

import java.io.Serializable;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.TimeUnit;

@ToString
public class SimpleCache<K extends Serializable & Comparable<K>, V extends Serializable> implements Cache<K, V> {
    private final ExpiringMap<K, V> internal;

    SimpleCache() {
        internal = ExpiringMap.builder()
                .expirationPolicy(ExpirationPolicy.CREATED)
                .variableExpiration()
                .build();
    }


    @Override
    public void put(K key, V value) {
        internal.put(key, value);
    }

    @Override
    public void put(K key, V value, long duration, TimeUnit timeUnit) {
        internal.put(key, value, duration, timeUnit);
    }

    @Override
    public void put(K key, V value, @NotNull Duration duration) {
        internal.put(key, value, duration.toNanos(), TimeUnit.NANOSECONDS);
    }

    @Override
    public boolean putIfAbsent(K key, V value) {
        return Objects.isNull(internal.putIfAbsent(key, value));
    }

    @Override
    public V getOrPut(K key, V value) {
        val v = internal.get(key);
        if (Objects.nonNull(v)) return v;
        return internal.put(key, value);
    }

    @Override
    public V getAndRemove(K key) {
        val value = internal.get(key);
        if (Objects.isNull(value)) return null;
        internal.remove(key);
        return value;
    }

    @Override
    public boolean hasKey(K key) {
        return internal.containsKey(key);
    }

    @Override
    public boolean hasValue(V value) {
        return internal.containsKey(value);
    }

    @Override
    public Optional<V> get(K key) {
        return Optional.of(internal.get(key));
    }

    @Override
    public @NotNull Set<K> getAllOfKey() {
        return internal.keySet();
    }

    @Override
    public @NotNull Collection<V> getAllOfValue() {
        return internal.values();
    }

    @Override
    public void remove(K key) {
        internal.remove(key);
    }

    @Override
    public void removeAll(@NotNull List<K> keys) {
        keys.stream().distinct().forEach(internal::remove);
    }

    @Override
    public void removeAll(@NotNull Set<K> keys) {
        keys.forEach(internal::remove);
    }

    @Override
    public void removeAll(K[] keys) {
        Arrays.stream(keys).forEach(internal::remove);
    }

    @Override
    public void clear() {
        internal.clear();
    }

    @Override
    public Optional<V> removeAndGet(K key) {
        val value = internal.get(key);
        internal.remove(key);
        return Optional.of(value);
    }

    @Override
    public void expire(K key, long t, TimeUnit unit) {
        internal.setExpiration(key, t, unit);
    }

    @Override
    public void expire(K key, @NotNull Duration duration) {
        internal.setExpiration(key, duration.toNanos(), TimeUnit.NANOSECONDS);
    }

    @Override
    public int size() {
        return internal.size();
    }

    @Override
    public int compareTo(@NotNull K o) {
        return o.hashCode();
    }

    @NotNull
    @Override
    public Iterator<Entry<K, V>> iterator() {
        return new Iterator<>() {
            @Override
            public boolean hasNext() {
                return internal.keySet().iterator().hasNext();
            }

            @Override
            public Entry<K, V> next() {
//                return internal.get(internal.keySet().iterator().next());
//                return iterator().next();
                return null;
            }
        };
    }
}

File: libs/cache/cache-simple/src/main/kotlin/org/toolkit4J/libs/cache/simple/SimpleCache.kt
package org.toolkit4J.libs.cache.simple

import net.jodah.expiringmap.ExpiringMap
import javax.cache.Cache
import javax.cache.Cache.Entry
import javax.cache.CacheManager
import javax.cache.configuration.CacheEntryListenerConfiguration
import javax.cache.configuration.Configuration
import javax.cache.integration.CompletionListener
import javax.cache.processor.EntryProcessor
import javax.cache.processor.EntryProcessorResult

class SimpleCache<K, V>(private val name: String) : Cache<K, V> {
    private var store = ExpiringMap.builder()
        .variableExpiration()
        .build<K, V>()

    override fun iterator(): MutableIterator<Entry<K, V>> {
//        store.entries.map { mutableEntry ->  {
//           val entry:Entry<K,V> = object :Entry<K,V>{
//               override fun getKey(): K {
//                  return mutableEntry.key
//               }
//
//               override fun getValue(): V {
//                   return mutableEntry.value
//               }
//
//               override fun <T : Any?> unwrap(clazz: Class<T>?): T {
//                   TODO("Not yet implemented")
//               }
//           }
//        }}.m.iterator();
        TODO("Not yet implemented")
    }

    override fun close() {
        store = null
    }

    override fun removeAll() {
        clear()
    }

    override fun clear() {
        store.clear()
    }

    override fun getName(): String {
        return name;
    }

    override fun getCacheManager(): CacheManager {
        TODO("Not yet implemented")
    }

    override fun isClosed(): Boolean {
        return store == null
    }

    override fun <T : Any?> unwrap(clazz: Class<T>?): T {
        TODO("Not yet implemented")
    }

    override fun deregisterCacheEntryListener(cacheEntryListenerConfiguration: CacheEntryListenerConfiguration<K, V>?) {
        TODO("Not yet implemented")
    }

    override fun registerCacheEntryListener(cacheEntryListenerConfiguration: CacheEntryListenerConfiguration<K, V>?) {
        TODO("Not yet implemented")
    }

    override fun <T : Any?> invokeAll(
        keys: MutableSet<out K>?,
        entryProcessor: EntryProcessor<K, V, T>?,
        vararg arguments: Any?
    ): MutableMap<K, EntryProcessorResult<T>> {
        TODO("Not yet implemented")
    }

    override fun <T : Any?> invoke(key: K, entryProcessor: EntryProcessor<K, V, T>?, vararg arguments: Any?): T {
        TODO("Not yet implemented")
    }

    override fun <C : Configuration<K, V>?> getConfiguration(clazz: Class<C>?): C {
        TODO("Not yet implemented")
    }

    override fun removeAll(keys: MutableSet<out K>?) {
        keys?.forEach { k: K -> store.remove(k) }
    }

    override fun getAndReplace(key: K, value: V): V {
        TODO("Not yet implemented")
    }

    override fun replace(key: K, value: V): Boolean {
        return store.replace(key, value) == null
    }

    override fun replace(key: K, oldValue: V, newValue: V): Boolean {
        TODO("Not yet implemented")
    }

    override fun getAndRemove(key: K): V {
        return store.remove(key)!!
    }

    override fun remove(key: K, oldValue: V): Boolean {
        return when (oldValue) {
            store.get(key) -> store.remove(key) == null
            else -> false
        }
    }

    override fun remove(key: K): Boolean {
        return store.remove(key) == null
    }

    override fun putIfAbsent(key: K, value: V): Boolean {
        return store.putIfAbsent(key, value) == null
    }

    override fun putAll(map: MutableMap<out K, out V>?) {
        map?.let { store.putAll(it) }
    }

    override fun getAndPut(key: K, value: V): V {
        TODO("Not yet implemented")
    }

    override fun put(key: K, value: V) {
        store.put(key, value)
    }

    override fun loadAll(
        keys: MutableSet<out K>?,
        replaceExistingValues: Boolean,
        completionListener: CompletionListener?
    ) {
        TODO("Not yet implemented")
    }

    override fun containsKey(key: K): Boolean {
        return store.containsKey(key)
    }

    override fun getAll(keys: MutableSet<out K>?): MutableMap<K, V> {
        return store.filterKeys { k: K -> keys!!.contains(k) }.toMutableMap()
    }

    override fun get(key: K): V {
        return store.get(key)!!
    }
}